From 12aeedaeea35223fb3c9684ae84576ed476f7979 Mon Sep 17 00:00:00 2001
From: Xinhui Yang <cyan@cyano.uk>
Date: Sat, 28 Dec 2024 16:06:22 +0800
Subject: [PATCH 4/4] exec-credential: try forking again without CLONE_NEWNS

QEMU user emulation currently does not support CLONE_NEWS, fork()ing
with it will result in an EINVAL, rendering the system unbootable with
systemd-nspawn.
---
 src/core/exec-credential.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/core/exec-credential.c b/src/core/exec-credential.c
index 6157ac4255..636092e9dc 100644
--- a/src/core/exec-credential.c
+++ b/src/core/exec-credential.c
@@ -976,6 +976,13 @@ int exec_setup_credentials(
                 return r;
 
         r = safe_fork("(sd-mkdcreds)", FORK_DEATHSIG_SIGTERM|FORK_WAIT|FORK_NEW_MOUNTNS, NULL);
+        if (r == -EINVAL) {
+                /* Well, try again without CLONE_NEWNS.
+                 * There's a chance that we might be running with QEMU user emulation.
+                 * QEMU user emulation does not support CLONE_NEWNS:
+                 * https://gitlab.com/qemu-project/qemu/-/issues/1962 */
+                r = safe_fork("(sd-mkdcreds)", FORK_DEATHSIG_SIGTERM|FORK_WAIT, NULL);
+        }
         if (r < 0) {
                 _cleanup_(rmdir_and_freep) char *u = NULL; /* remove the temporary workspace if we can */
                 _cleanup_free_ char *t = NULL;
-- 
2.47.1

